IO/d4est_norms.c

```diff
@@ -0,0 +1,617 @@
+(
+  const char** field_names,
+  const char** norm_names
+){
+  int i_fields = -1;
+  while (field_names[++i_fields] != NULL) {
+
+    char *norms_output_category;
+    asprintf(&norms_output_category, "norms_%s", field_names[i_fields]);
+    zlog_category_t *c_norms_output = zlog_get_category(norms_output_category);
+    free(norms_output_category);
+    
+    char *header;
+    asprintf(&header, "num_quadrants num_nodes avg_deg num_quad_nodes avg_deg_quad");
+    int i_norms = -1;
+    while (norm_names[++i_norms] != NULL) {
+      asprintf(&header, "%s %s",header, norm_names[i_norms]);
+    }
+    zlog_info(c_norms_output, header);
+  }
+}
+
+
+static void
+d4est_norms_calculate_analytic_error // TODO: remove with d4est_norms_norms_using_analytic_solution
+(
+ p4est_t* p4est,
+ d4est_operators_t* d4est_ops,
+ d4est_geometry_t* d4est_geom,
+ d4est_quadrature_t* d4est_quad,
+ double* u,
+ int local_nodes,
+ d4est_xyz_fcn_t analytic_solution,
+ void* analytic_solution_ctx,
+ double* u_analytic,
+ double* error
+)
+{
+  d4est_mesh_init_field
+    (
+     p4est,
+     u_analytic,
+     analytic_solution,
+     d4est_ops,
+     d4est_geom,
+     INIT_FIELD_ON_LOBATTO,
+     analytic_solution_ctx
+    );
+
+  /* for (int i = 0; i < prob_vecs->local_nodes; i++) */
+  /* printf("u_analytic[%d] = %.25f\n",i, u_analytic[i]); */
+    
+  d4est_linalg_vec_axpyeqz(-1., u, u_analytic, error, local_nodes);
+  d4est_linalg_vec_fabs(error, local_nodes);
+}
+
+void
+d4est_norms_energy_norm_fit
+(
+ p4est_t* p4est,
+ d4est_norms_energy_norm_fit_t* fit
+)
+{
+  double slope;
+  double intercept;
+  d4est_util_linear_regression
+    (
+     fit->log_energy_norm_data,
+     fit->dof_data,
+     &slope,
+     &intercept,
+     fit->stride
+    );
+
+  if (p4est->mpirank == 0) {
+    zlog_category_t *c_default = zlog_get_category("d4est_norms_energy_fit");
+    zlog_info(c_default, "(2): C1 = %f, C2 = %.15f", intercept, slope);
+  }
+}
+
+void
+d4est_norms_energy_norm_add_entry_and_fit
+(
+ p4est_t* p4est,
+ d4est_norms_energy_norm_fit_t* fit,
+ double global_energy_norm_sqr,
+ double global_dof
+){
+  if (global_energy_norm_sqr > 0.){
+    fit->log_energy_norm_data[fit->stride] = log(sqrt(global_energy_norm_sqr));
+    fit->dof_data[fit->stride] = pow(global_dof, 1./(2.*(P4EST_DIM)-1.));
+    fit->stride += 1;
+
+    if (fit->stride >= 2){
+      if (p4est->mpirank == 0){
+        zlog_category_t *c_default = zlog_get_category("d4est_norms_energy_fit");
+        zlog_info(c_default, "(1): ||err|| = C1*exp(C2*DOF^(1/%d))", 2*(P4EST_DIM)-1);
+        zlog_info(c_default, "(1): ||err|| = %.15f", sqrt(global_energy_norm_sqr));
+      }
+      d4est_norms_energy_norm_fit(p4est,fit);
+    }
+  }
+}
+
+void
+d4est_norms_destroy_energy_norm_fit
+(
+ d4est_norms_energy_norm_fit_t* fit
+)
+{
+  P4EST_FREE(fit->log_energy_norm_data);
+  P4EST_FREE(fit->dof_data);
+  P4EST_FREE(fit);
+}
+
+d4est_norms_norms_t
+d4est_norms_norms // TODO: Remove, superseded by `d4est_norms_save`
```

Comment:
We can remove this function as soon as it is not used anymore, i.e. we can gradually it with `d4est_norms_save` (or all at once at some point). With it, `d4est_norms_norms_t`, `d4est_norms_norms_using_analytic_solution` and `d4est_norms_write_header` can go as well.

Reply:
Yes anything else besides d4est_norms_save can just be deleted because d4est_norms_save has all the functionality.